restapi/src/test/java/io/stargate/web/docsapi/service/query/ExpressionParserIntTest.java

```diff
@@ -131,17 +132,114 @@ public void andWrappingRespected() {
                                     }));
               });
     }
+
+    @Test
+    public void complex() throws Exception {
+      // a & (b | (c & d))
+      String json =
+          "{\"a\": {\"$eq\": \"a\"}, \"$or\": [{\"b\": {\"$eq\": \"b\"}}, {\"$and\":[{\"c\": {\"$eq\": \"c\"}},{\"d\": {\"$eq\": \"d\"}}]}]}";
+      JsonNode root = mapper.readTree(json);
+
+      Expression<FilterExpression> result =
+          service.constructFilterExpression(Collections.emptyList(), root, false);
+
+      assertThat(result)
+          .isInstanceOfSatisfying(
+              And.class,
+              a -> {
+                List<?> children = a.getChildren();
+                assertThat(children)
+                    .hasSize(2)
+                    .anySatisfy(
+                        c ->
+                            assertThat(c)
+                                .isInstanceOfSatisfying(
+                                    FilterExpression.class,
+                                    first -> {
+                                      assertThat(first.getOrderIndex()).isZero();
+                                      assertThat(first.getFilterPath().getField()).isEqualTo("a");
+                                    }))
```

Comment:
Should we assert this on the _first_ element, rather than "any"?

Reply:
No, the order is not important and will be defined by the stupid hash comparator. What's important is that `orderIndex=0` is there for the field `a`.

Reply:
right, sorry about the false alarm :facepalm: